crates/scmi/src/main.rs

```diff
@@ -0,0 +1,213 @@
+//! which is done using `--device` command line option.  It can be used more
+//! than once, for different devices.  `--help-devices` lists the available
+//! devices and their options.
+//!
+//! The daemon normally logs info and higher messages to the standard error
+//! output.  To log more messages, you can set `RUST_LOG` environment variable,
+//! e.g. to `debug`.
+//!
+//! Here is an example command line invocation of the daemon:
+//!
+//! ```sh
+//! RUST_LOG=debug vhost-device-scmi \
+//!   --socket ~/tmp/scmi.sock \
+//!   --device iio,path=/sys/bus/iio/devices/iio:device0,channel=in_accel
+//! ```
+
+mod devices;
+mod scmi;
+mod vhu_scmi;
+
+use std::{
+    println,
+    sync::{Arc, RwLock},
+};
+
+use clap::{CommandFactory, Parser};
+use devices::common::{print_devices_help, DeviceDescription, DeviceProperties};
+use itertools::Itertools;
+use log::{debug, info, warn};
+
+use vhost::vhost_user;
+use vhost::vhost_user::Listener;
+use vhost_user_backend::VhostUserDaemon;
+use vhu_scmi::VuScmiBackend;
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+type Result<T> = std::result::Result<T, String>;
+
+#[derive(Parser)]
+struct ScmiArgs {
+    // Location of vhost-user Unix domain socket.
+    // Required, unless one of the --help options is used.
+    #[clap(short, long, help = "vhost-user socket to use (required)")]
+    socket_path: Option<String>,
+    // Specification of SCMI devices to create.
+    #[clap(short, long, help = "Devices to expose")]
+    #[arg(num_args(1..))]
+    device: Vec<String>,
+    #[clap(long, help = "Print help on available devices")]
+    help_devices: bool,
+}
+
+struct VuScmiConfig {
+    socket_path: String,
+    devices: DeviceDescription,
+}
+
+impl TryFrom<ScmiArgs> for VuScmiConfig {
+    type Error = String;
+
+    fn try_from(cmd_args: ScmiArgs) -> Result<Self> {
+        if cmd_args.socket_path.is_none() {
+            return Result::Err("Required argument socket-path was not provided".to_string());
+        }
+        let socket_path = cmd_args.socket_path.unwrap().trim().to_string();
+        let mut devices: DeviceDescription = vec![];
+        let device_iterator = cmd_args.device.iter();
+        for d in device_iterator {
+            let mut split = d.split(',');
+            let name = String::from(split.next().unwrap());
+            let mut properties = vec![];
+            for s in split {
+                if let Some((key, value)) = s.split('=').collect_tuple() {
+                    properties.push((String::from(key), String::from(value)));
+                } else {
+                    return Result::Err(format!("Invalid device {name} property format: {s}"));
+                }
+            }
+            devices.push((name, DeviceProperties::new(properties)));
+        }
+        Ok(VuScmiConfig {
+            socket_path,
+            devices,
+        })
+    }
+}
+
+fn start_backend(config: VuScmiConfig) {
```

Comment:
Maybe this should return Result<(), Error> and print the error later?

Reply:
Yes, done.